File: 2017/cpp/day01.cpp
#include <iostream>
using namespace std;

int part1(string s) {
  int total = 0;
  auto last = s.back();

  for (auto it = s.begin(); it != s.end(); it++) {
    if (*it == last) {
      total += (*it) - '0';
    }

    last = *it;
  }

  return total;
}

int part2(string s) {
  string s2 = s.substr(s.size() / 2) + s.substr(0, s.size() / 2);
  int total = 0;
  int len = s.size();

  for (int i = 0; i < len; i++) {
    if (s[i] == s2[i]) {
      total += s[i] - '0';
    }
  }

  return total;
}

int main() {
  string s;
  cin >> s;

  cout << "Part 1: " << part1(s) << "\n";
  cout << "Part 2: " << part2(s) << "\n";
}

File: 2017/rust/src/bin/day01.rs
use std::io::{self, Read};

fn main() {
    let mut s = String::new();
    io::stdin().read_to_string(&mut s).unwrap();

    let v: Vec<u32> = s.trim()
        .chars()
        .map(|c| c.to_digit(10).unwrap()).collect();

    println!("Part 1: {}", part1(&v));
    println!("Part 2: {}", part2(&v));
}

fn part1(v: &[u32]) -> u32 {
    let mut last = v.last().unwrap();
    let mut total = 0;
    for n in v {
        if n == last {
            total += n;
        }
        last = n;
    }
    total
}

fn part2(v: &[u32]) -> u32 {
    let i1 = v.iter();
    let i2 = v.iter().cycle().skip(v.len() / 2);

    i1.zip(i2).filter(|(a, b)| a == b).map(|(a, _)| a).sum()
}
